src/Compilers/CSharp/Test/Semantic/Semantics/RefExtensionMethodsTests.cs

```diff
@@ -0,0 +1,1151 @@
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
+{
+    public class RefExtensionMethodsTests : CSharpTestBase
+    {
+        [Fact]
+        public void ExtensionMethods_RValues_Ref_NotAllowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref this int p)
+    {
+        System.Console.WriteLine(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        5.PrintValue();
+    }
+}";
+
+            CreateCompilationWithMscorlibAndSystemCore(code).VerifyDiagnostics(
+                // (13,9): error CS1510: A ref or out value must be an assignable variable
+                //         5.PrintValue();
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "5").WithLocation(13, 9));
+        }
+
+        [Fact]
+        public void ExtensionMethods_LValues_Ref_Allowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref this int p)
+    {
+        System.Console.Write(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        int x = 5;
+        x.PrintValue();
+        Extensions.PrintValue(ref x);
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "55");
+        }
+
+        [Fact]
+        public void ExtensionMethods_RValues_RefReadOnly_Allowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref readonly this int p)
+    {
+        System.Console.Write(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        5.PrintValue();
+        Extensions.PrintValue(5);
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "55");
+        }
+
+        [Fact]
+        public void ExtensionMethods_LValues_RefReadOnly_Allowed()
+        {
+            var code = @"
+public static class Extensions
+{
+    public static void PrintValue(ref readonly this int p)
+    {
+        System.Console.Write(p);
+    }
+}
+public static class Program
+{
+    public static void Main()
+    {
+        int x = 5;
+        x.PrintValue();
+        Extensions.PrintValue(x);
+    }
+}";
+
+            CompileAndVerify(code, additionalRefs: new[] { SystemCoreRef }, expectedOutput: "55");
+        }
+
+        [Fact]
+        public void ExtensionMethods_NullConditionalOperator_Ref_NotAllowed()
+        {
+            var code = @"
+public struct TestType
+{
+    public int GetValue() => 0;
+}
+public static class Extensions
+{
+    public static void Call(ref this TestType obj)
+    {
+        var value1 = obj?.GetValue();        // This should be an error
```

Comment:
>var value1 = obj?.GetValue();        // This should be an error [](start = 8, length = 63)

We should also test a scenario when ```this``` parameter is not a struct. An extension method like that can be defined in VB or IL.